### PR TITLE
Implement uploading by chunks in fileUploader

### PR DESCRIPTION
**Pull request by andrewmakarov**

### CODE REVIEW

File: js/ui/file_uploader.js [hunk 1]
```diff
@@ -1119,16 +1125,125 @@ var FileUploader = Editor.inherit({
...
+        if(file.slice) {
+            return file.slice(startPos, startPos + length);
+        }
+        if(file.webkitSlice) {
+            return file.webkitSlice(startPos, startPos + length);
+        }
+        if(file.mozSlice) {
+            return file.mozSlice(startPos, startPos + length);
+        }
+        return null;
```
Comment (jsdmitry) [inline js/ui/file_uploader.js]: Maybe to make easier:
```
var slice = file.size || file.webkitSlice || file.mozSlice;
return typeUtils.isDefined(slice) ? slice(startPos, startPos + length) : null;

```

File: js/ui/file_uploader.js [hunk 2]
```diff
@@ -1119,16 +1125,125 @@ var FileUploader = Editor.inherit({
...
+    _createChunkFormData: function(fileName, blobName, blob, index, count, type, guid, size) {
+        var formData = new window.FormData();
+        formData.append(blobName, blob);
+        formData.append("metaData", JSON.stringify({
+            Name: fileName,
+            Index: index,
+            Count: count,
+            FileSize: size,
+            Type: type,
+            Guid: guid
```
Comment (jsdmitry) [inline js/ui/file_uploader.js]: Why uses the upper case for the name of fields?

File: js/ui/file_uploader.js [hunk 3]
```diff
@@ -1119,16 +1125,125 @@ var FileUploader = Editor.inherit({
...
-            file.progressBar.$element().appendTo($file);
+    _sliceFile: function(file, startPos, length) {
+        if(file.slice) {
+            return file.slice(startPos, startPos + length);
+        }
+        if(file.webkitSlice) {
+            return file.webkitSlice(startPos, startPos + length);
+        }
+        if(file.mozSlice) {
+            return file.mozSlice(startPos, startPos + length);
```
Comment (DokaRus) [inline js/ui/file_uploader.js]: `mozSlice` is actual for FireFox 12 and earlier.
https://developer.mozilla.org/en-US/docs/Web/API/Blob/slice

File: js/ui/file_uploader.js [hunk 4]
```diff
@@ -1597,6 +1363,237 @@ FileUploader.__internals = {
 };
 ///#ENDDEBUG
 
+var FileUploadStrategyBase = Class.inherit({
+    ctor: function(fileUploader) {
+        this.fileUploader = fileUploader;
+    },
+    upload: function(file) {
```
Comment (jsdmitry) [inline js/ui/file_uploader.js]: Add an empty line after the methods. For example, please see the button.js file.

File: js/ui/file_uploader.js (hunk 1 above)
Comment (andrewmakarov) [reply js/ui/file_uploader.js]: This approach will not work.
In this case, it is necessary to write: `slice.call(file, startPos, startPos + length)` - looks not pretty

File: js/ui/file_uploader.js [hunk 5]
```diff
@@ -1597,6 +1363,237 @@ FileUploader.__internals = {
...
+    _sliceFile: function(file, startPos, length) {
+        if(file.slice) {
+            return file.slice(startPos, startPos + length);
+        }
+        if(file.webkitSlice) {
+            return file.webkitSlice(startPos, startPos + length);
+        }
+        return null;
+    },
+    _getEvent: function(e) {
```
Comment (jsdmitry) [inline js/ui/file_uploader.js]: Why need the "_getEvent" method?

Comment (jsdmitry) [reply js/ui/file_uploader.js]: Maybe to set "e" to "event" always inside the base strategy.

File: js/ui/file_uploader.js [hunk 6]
```diff
@@ -1481,6 +1363,252 @@ FileUploader.__internals = {
...
+                if(this._isStatusError(e.status)) {
+                    file._isError = true;
+                    file.onError.fire();
+                }
+                chunksData.chunks = [];
+            }.bind(this));
+        }
+    },
+
+    _createFormData: function(fileName, blobName, blob, index, count, type, guid, size) {
```
Comment (jsdmitry) [inline js/ui/file_uploader.js]: Use one input parameter - "options" instead of big count of parameters

File: js/ui/file_uploader.js (hunk 5 above)
Comment (andrewmakarov) [reply js/ui/file_uploader.js]: In whole file uploading mode, to an onProgress event is passed event from the ajax request.
In the case of uploading by chunks, each chunk has own request, so it's useless to forward it to an onProgress event.

Comment (andrewmakarov) [reply js/ui/file_uploader.js]: Probably, to have a single style in the API, it makes sense to pass the request event

File: js/ui/file_uploader.js [hunk 7]
```diff
@@ -1481,6 +1363,252 @@ FileUploader.__internals = {
...
+    },
+
+    _sliceFile: function(file, startPos, length) {
+        if(file.slice) {
+            return file.slice(startPos, startPos + length);
+        }
+        if(file.webkitSlice) {
+            return file.webkitSlice(startPos, startPos + length);
+        }
+        return null;
```
Comment (jsdmitry) [inline js/ui/file_uploader.js:1562]: Why do not like?:
```
var slice = file.slice || file.webkitSlice;
return  slice ? slice(startPos, startPos + length) : null;
```

Comment (andrewmakarov) [reply js/ui/file_uploader.js:1562]: This approach will not work.
In this case, it is necessary to write: `slice.call(file, startPos, startPos + length)` - looks not pretty

**Review by jsdmitry — APPROVED** [review]

**Pull request merged**